### DISCUSSION TITLE
Questions Regarding SMI Metrics

### DISCUSSION THREAD

**Discussion by snehachhabria** [servicemeshinterface/smi-spec#165] [General]
1. Does The TrafficMetrics CRD in SMI metrics in any way control the proxy level configurations (i.e what metrics to scrape from the proxy for the resource or does it control Prometheus configurations) ? Or is it only used to control the requests made to the Kubernetes API server via the SMI Metrics API 
2. What’s the purpose of TrafficMetricsList ? My understanding is its just a way to provide a list of resources rather than a single one in the TrafficMetrics CRD
3. The SMI implementation of Metrics would require the application of a TrafficMetric and/or TrafficMetricsList CRD and the APIService CDR or this there any other CRD ? 
4. If I have a TrafficMetric CRD defined only for a resource of type Service and some user makes a request for another resource like deployment, how does /apis/metrics.smi-spec.io/v1alpha1/ handle this ?

**Comment by grampelberg** [comment]
These are great questions! Lemme try and tackle them inline.

> Does The TrafficMetrics CRD in SMI metrics in any way control the proxy level configurations (i.e what metrics to scrape from the proxy for the resource or does it control Prometheus configurations) ? Or is it only used to control the requests made to the Kubernetes API server via the SMI Metrics API

Nope, `TrafficMetrics` isn't a CRD. It is actually an [APIService](https://kubernetes.io/docs/tasks/access-kubernetes-api/setup-extension-api-server/) that uses the [aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) and works identically to [metrics.k8s.io](https://kubernetes.io/docs/tasks/debug-application-cluster/resource-metrics-pipeline/#the-metrics-api).

So, all the configuration happens in the backend service itself (your implementation). If you take a look at the [smi-metrics code](https://github.com/servicemeshinterface/smi-metrics/blob/master/pkg/metrics/handler.go#L29), you'll see where that's happening on a per-handler basis right now.

> What’s the purpose of TrafficMetricsList ? My understanding is its just a way to provide a list of resources rather than a single one in the TrafficMetrics CRD

Yup! Same as [PodList](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podlist-v1-core) for example.

> The SMI implementation of Metrics would require the application of a TrafficMetric and/or TrafficMetricsList CRD and the APIService CDR or this there any other CRD ?

You'd need the `APIService` definition and a service running to handle the requests.

> If I have a TrafficMetric CRD defined only for a resource of type Service and some user makes a request for another resource like deployment, how does /apis/metrics.smi-spec.io/v1alpha1/ handle this ?

`TrafficMetrics` don't currently support `Service` types (for long winded reasons). That said, it is up to your implementation to make decisions about what you'd like to surface. An implementation could technically only report on actual resources like `Pod` instead of owners like `Deployment`.

**Comment by michelleN** [comment]
Hey @grampelberg - The traffic metrics spec states that it does support services.

See https://github.com/servicemeshinterface/smi-spec/blob/master/traffic-metrics.md#kubernetes-api
```
The full list of resources for this list would be:

namespaces
nodes
pods
replicationcontrollers
services
daemonsets
deployments
replicasets
statefulsets
jobs
trafficsplits
```

Is this list just incorrect?

**Reply by grampelberg** [reply]
That's what I get for trying to be over-explicit. Services are weird in that you can only get some metrics for them (basically only `to`). I thought I'd left them out because, unlike the other resources, they come with a caveat. It is fine to leave them in, or take them out. Either way, probably worth a discussion if we're running into issues.

**Comment by snehachhabria** [comment]
> Nope, `TrafficMetrics` isn't a CRD. It is actually an [APIService](https://kubernetes.io/docs/tasks/access-kubernetes-api/setup-extension-api-server/) that uses the [aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) and works identically to [metrics.k8s.io](https://kubernetes.io/docs/tasks/debug-application-cluster/resource-metrics-pipeline/#the-metrics-api).
> 
> So, all the configuration happens in the backend service itself (your implementation). If you take a look at the [smi-metrics code](https://github.com/servicemeshinterface/smi-metrics/blob/master/pkg/metrics/handler.go#L29), you'll see where that's happening on a per-handler basis right now.
> 
Thanks for clarifying this, it really helps in understanding the purpose of the SMI metrics
> `TrafficMetrics` don't currently support `Service` types (for long winded reasons). That said, it is up to your implementation to make decisions about what you'd like to surface. An implementation could technically only report on actual resources like `Pod` instead of owners like `Deployment`.

Would you be able to shed more light on the types that are currently supported, if the full list that @michelleN  shared isn’t applicable

**Reply by grampelberg** [reply]
I was mistaken, sorry about that! You can't do the full set of edges with `Service`, but it is possible to get out. It is probably worth a longer conversation as to whether we're happy with the tradeoffs or not.
